pkg/controller/container-runtime-config/container_runtime_config_controller_test.go

```diff
@@ -830,30 +832,88 @@ func TestRegistriesValidation(t *testing.T) {
 				InsecureRegistries: []string{"test.io"},
 			},
 		},
+		{
+			name: "adding registry used by payload to blocked registries with mirror rules configured for the a different repo in the reg",
+			config: &apicfgv1.RegistrySources{
+				BlockedRegistries:  []string{"blah.io", "docker.io"},
+				InsecureRegistries: []string{"test.io"},
+			},
+			icspRules: []*apioperatorsv1alpha1.ImageContentSourcePolicy{
+				{
+					Spec: apioperatorsv1alpha1.ImageContentSourcePolicySpec{
+						RepositoryDigestMirrors: []apioperatorsv1alpha1.RepositoryDigestMirrors{
+							{Source: "blah.io/myuser", Mirrors: []string{"mirror-1.io/myuser", "mirror-2.io/myuser"}},
+						},
+					},
+				},
+			},
+		},
 	}
 
 	successTests := []struct {
-		name   string
-		config *apicfgv1.RegistrySources
+		name                                                              string
+		expectedRegistriesBlocked, expectedPolicyBlocked, expectedAllowed []string
+		config                                                            *apicfgv1.RegistrySources
+		icspRules                                                         []*apioperatorsv1alpha1.ImageContentSourcePolicy
 	}{
 		{
-			name: "adding registry used by payload to blocked registries",
+			name: "adding registry used by payload to insecure registries",
 			config: &apicfgv1.RegistrySources{
 				BlockedRegistries:  []string{"docker.io"},
 				InsecureRegistries: []string{"blah.io"},
 			},
+			expectedRegistriesBlocked: []string{"docker.io"},
+			expectedPolicyBlocked:     []string{"docker.io"},
+		},
+		{
+			name: "adding registry used by payload to blocked registries with mirror rules configured for the payload reg",
+			config: &apicfgv1.RegistrySources{
+				BlockedRegistries:  []string{"blah.io", "docker.io"},
+				InsecureRegistries: []string{"test.io"},
+			},
+			icspRules: []*apioperatorsv1alpha1.ImageContentSourcePolicy{
+				{
+					Spec: apioperatorsv1alpha1.ImageContentSourcePolicySpec{
+						RepositoryDigestMirrors: []apioperatorsv1alpha1.RepositoryDigestMirrors{
+							{Source: "blah.io/payload", Mirrors: []string{"mirror-1.io/payload", "mirror-2.io/payload"}},
+						},
+					},
+				},
+			},
+			expectedRegistriesBlocked: []string{"blah.io", "docker.io"},
+			expectedPolicyBlocked:     []string{"blah.io", "docker.io"},
+			expectedAllowed:           []string{"blah.io/payload/myimage"},
+		},
+		{
+			name: "adding payload repository to blocked registries with mirror rules configured for the payload registry",
+			config: &apicfgv1.RegistrySources{
+				BlockedRegistries:  []string{"blah.io/payload", "docker.io"},
+				InsecureRegistries: []string{"test.io"},
+			},
+			icspRules: []*apioperatorsv1alpha1.ImageContentSourcePolicy{
+				{
+					Spec: apioperatorsv1alpha1.ImageContentSourcePolicySpec{
+						RepositoryDigestMirrors: []apioperatorsv1alpha1.RepositoryDigestMirrors{
+							{Source: "blah.io", Mirrors: []string{"mirror-1.io", "mirror-2.io"}},
+						},
+					},
+				},
+			},
```

Comment:
Oh… this is a problematic case.

`registries.conf` only uses the `sysregistriesv2.Registry` entry with the _closest matching_ prefix. So, `BlockedRegistries: blah.io/payload` creates a `Registry` entry which “shadows” the `blah.io` entry for ICSP, effectively completely disabling the mirror configuration for the `blah.io` namespace.

OTOH OpenShift’s `BlockedRegistries` and ICSP are not defined to interact in any way. That’s a bug in runtime-utils: it can propagate blocked/insecure to nested sub-scopes, but it can’t propagate mirrors to nested sub-scopes.

For the purpose of _this_ PR, I think changing the three scopes in the mirror to add `/payload` (resulting in one `Registry` entry that carries both the blocked flat and the mirrors) would work. Separately, runtime-utils will need fixing…

Reply:
… Actually… if `blah.io/$repo` is blocked in registries.conf, it it _also_ blocked in `policy.json`, so the set of mirrors configured for blah.io/$repo` currently doesn’t matter at all. The only current exception is payloadRepo itself, which we treat specially. So this bug is not nearly that severe. (That will change with ITMS/IDMS and their `neverContactSource` option.)

Reply:
… correcting myself one more time, even with `neverContactSource` this is not relevant, because `neverContactSource` would set up the mirror and the blocked bit on the _same_ `Registry` object. If there is a mirror configuration for $superScope, then any $subScope is either configuring a different mirror set (= no need to inherit the mirrors), or it is just adding a blocked bit (= no point in inheriting the mirrors).

So… this really only matters for, basically `BlockedRegistries: quay.io`, doesn’t it? But in that case we do need to propagate the mirrors… Let me take a stab at that.

Reply:
https://github.com/openshift/runtime-utils/pull/16 should help.

Reply:
Yeah, I started hitting this when I fixed up the code based on comments. I am still a little confused why this is happening, basically `PullSourcesFromReference` is unable to detect that `mirror-1.io/payload/image` is a mirror of the release image. 
So what exactly in runtime-utils needs fixing for this?

I have added `/payload` to fix the test for now.

Reply:
The test caused generation of `registries.conf` roughly like this:
```toml
[[registry]]
location = "blah.io/payload"
blocked = true

[[registry]]
location = "blah.io"
mirrors = …
```

The way c/image works, only the most specific match matters, i.e. the `blah.io/payload` entry shadows the `blah.io` entry. So, given a config like this, `payloadRepoHasUnblockedMirror("blah.io/payload")` _correctly_ mirrors c/image behavior and finds no mirrors that apply to `blah.io`.

Does that help clarify, or is there something else unexpected going on that I’m missing?

I think https://github.com/openshift/runtime-utils/pull/16 fixes the underlying cause (of not generating `registries.conf` that correctly interprets the provided BlockedRegistries + ICSP), at least per its unit tests — but I didn’t actually go back to this test to verify that it now behaves as expected.

---

This problem manifests for configurations where blocking is nested inside mirrors, e.g. (`quay.io` has mirrors configured, `quay.io/openshift-release-dev` is blocked); those are technically legal, but I’d expect them to be much rarer than nesting mirrors inside blocking, e.g. (all of `quay.io` is blocked, `quay.io/openshift-release-dev` has mirrors).

So… I think this is a bug that we definitely need to fix, but I’m not at this point sure that it needs to block _this_ PR. (Also noting that I have had to re-evaluate my analysis two times, so it’s fairly likely I’m wrong again. I’d definitely appreciate a careful independent look.)

Reply:
> but I didn’t actually go back to this test to verify that it now behaves as expected.

Confirming that the runtime-utils PR allows this test, as it was originally, to pass.

Reply:
Awesome! I will pull it in here and fix up the test case once it is merged.